### ISSUE TITLE
ADFS Assertion Compatibility

### ISSUE THREAD

**Issue by drelyn86** [kislyuk/signxml#6]
The ADFS server I am working with is sending a reference URI like "#_639146ac-a51f-44a3-8a2d-c2c4d944c7d3". However, the element it is referring to is identified by an "AssertionID" attribute, instead of just "ID".

Here is the workaround I'm using right now:

    from signxml import xmldsig, InvalidInput

    class ADFSAssertionDSig(xmldsig):
    
        def _resolve_reference(self, doc_root, reference, uri_resolver=None):
            try:
                return xmldsig._resolve_reference(self, doc_root, reference, uri_resolver)
            except InvalidInput:
                return self._resolve_reference_by_assertion_id_attribute(doc_root, reference, uri_resolver)
    
        def _resolve_reference_by_assertion_id_attribute(self, doc_root, reference, uri_resolver=None):
            uri = reference.get("URI")
            results = doc_root.xpath("..//*[@AssertionId=$uri]", uri=uri.lstrip("#"))
            if len(results) < 1:
                results += doc_root.xpath("..//*[@AssertionID=$uri]", uri=uri.lstrip("#"))
            if len(results) < 1:
                raise InvalidInput("Unable to resolve reference URI: {}".format(uri))
            return results[0]

I'm thinking there should be a more elegant solution.

BTW, I thanks for the Python 3 fixes. I was actually working on a pull request last weekend for that, only to see on Monday that you had got it done.

**Comment by kislyuk** [comment]
Thanks for the heads up about this assertion format, I wasn't familiar with it. It will take a little while to sketch out the right API for this, if you have any suggestions, please post them.

**Comment by kislyuk** [comment]
@drelyn86, would you mind attaching or pointing me to (or sending through email) a complete assertion XML for an ADFS assertion?

**Comment by drelyn86** [comment]
@kislyuk sent to your gmail.

**Comment by kislyuk** [comment]
@drelyn86, please take a look at master. Does this API look all right to you?
```
with open("adfs.xml", "rb") as fh:
    d = etree.parse(fh).getroot()

s = d.find(".//{urn:oasis:names:tc:SAML:1.0:assertion}Assertion")
c = s.find(".//{http://www.w3.org/2000/09/xmldsig#}X509Certificate").text

xmldsig(s).verify(x509_cert=c, id_attribute="AssertionID")
```

(Please note as always, you should verify against a trusted CA cert obtained through a secure channel)

**Comment by drelyn86** [comment]
Yes, that works perfectly.

**Comment by kislyuk** [comment]
Released in v0.3.7.
